### PR TITLE
adding factoring support

### PR DESCRIPTION
**Pull request by aelnaiem**
Issue: https://github.com/socraticorg/mathsteps/issues/50

Adds support for factoring by:
* difference of squares
* perfect squares
* sum product rule

Quadratic equation to be added in the future. The reason I didn't add it yet is because it requires simplification of only what will be in the brackets (without expansion), and I didn't want to make a decision on how that should happen.

### CODE REVIEW

File: lib/node/Creator.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ const NodeCreator = {
     if (exponent) {
       polyTerm = this.operator('^', [polyTerm, exponent]);
     }
-    if (coeff) {
+    if (coeff && coeff.value !== '1') {
```
Comment (aelnaiem) [inline lib/node/Creator.js]: This breaks some tests. Considering just doing this check before calling this. What do you think @evykassirer

**Comment by evykassirer** [comment]
I think I'll get to this tomorrow :) this week is a bit busy for me and I think this isn't too urgent, ye?

**Comment by aelnaiem** [comment]
Not urgent

File: lib/checks/isQuadratic.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const Node = require('../node');
+
+// TODO(ael)
```
Comment (evykassirer) [inline lib/checks/isQuadratic.js]: docstring

File: lib/checks/isQuadratic.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+'use strict';
```
Comment (evykassirer) [inline lib/checks/isQuadratic.js]: can you merge in the new linter changes? apparently we don't need `'use strict'` in modules because they're strict by default (the linter will catch this now 😄  )

File: lib/checks/isQuadratic.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+const Node = require('../node');
+
+// TODO(ael)
+function isQuadratic(node) {
+  if (!Node.Type.isOperator(node) || node.op !== '+') {
```
Comment (evykassirer) [inline lib/checks/isQuadratic.js]: you can now do `if (!Node.Type.isOperator(node, '+')` thanks to a recent change :)

File: lib/checks/isQuadratic.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  if (node.args.length > 3) {
+    return false;
+  }
+
+  const secondDegreeTerms = node.args.filter(isSecondDegree);
+  const firstDegreeTerms = node.args.filter(isFirstDegree);
+  const constantTerms = node.args.filter(Node.Type.isConstant);
+
+  if (secondDegreeTerms.length !== 1 || firstDegreeTerms.length > 1 ||
+    constantTerms.length !== 1) {
```
Comment (evykassirer) [inline lib/checks/isQuadratic.js]: hmm why'd you choose this combination?

I agree that there should be a secondDegreeTerm for sure

but what about `x^2 + x`? that seems like a quadratic, but there's no constant term

File: lib/checks/isQuadratic.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    return false;
+  }
+
+  if (node.args.length > 3) {
+    return false;
+  }
+
+  const secondDegreeTerms = node.args.filter(isSecondDegree);
+  const firstDegreeTerms = node.args.filter(isFirstDegree);
+  const constantTerms = node.args.filter(Node.Type.isConstant);
```
Comment (evykassirer) [inline lib/checks/isQuadratic.js]: what if there are other args that don't fall into one of these buckets? we should return false right? so I guess check that the length of each of these add up to the number of args

File: lib/checks/isQuadratic.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+function isFirstDegree(node) {
+  if (Node.PolynomialTerm.isPolynomialTerm(node)) {
+    const polyTerm = new Node.PolynomialTerm(node);
+    const exponent = polyTerm.getExponentNode(true);
+    return exponent && exponent.value === '1';
+  }
+  return false;
+}
+
+function isSecondDegree(node) {
```
Comment (evykassirer) [inline lib/checks/isQuadratic.js]: could you just have a general function `isPolynomialTermOfDegree(node, degree)` to avoid duplicated code?

File: test/checks/isQuadratic.test.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+const assert = require('assert');
+const math = require('mathjs');
+
+const flatten = require('../../lib/util/flattenOperands');
+const checks = require('../../lib/checks');
+
+function testIsQuadratic(exprString, result) {
```
Comment (evykassirer) [inline test/checks/isQuadratic.test.js]: I just merged an update to add `testUtil`, so these tests should be of the new format now

File: test/checks/isQuadratic.test.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    ['x^2 - 4', true],
+    ['x^2 + 2x + 1', true],
+    ['x^2 - 2x + 1', true],
+    ['x^2 + 3x + 2', true],
+    ['x^2 - 3x + 2', true],
+    ['x^2 + x - 2', true],
+    ['x^2 + 4', true],
+    ['x^2 + 4x + 1', true],
+    ['x^2', false],
+    ['x^3 + x^2 + x + 1', false],
```
Comment (evykassirer) [inline test/checks/isQuadratic.test.js]: add tests for `x^2 + x` and `x^2 + 4 + 2^x` and other edge cases I brought up

File: lib/factor/ConstantFactors.js [hunk 10]
```diff
@@ -43,8 +43,11 @@ ConstantFactors.getPrimeFactors = function(number){
 ConstantFactors.getFactorPairs = function(number){
   const factors = [];
 
-  const root = Math.sqrt(number);
-  for (var divisor = 1; divisor <= root; divisor++) {
+  const bound = Math.floor(Math.sqrt(Math.abs(number)));
+  for (var divisor = -bound; divisor <= bound; divisor++) {
```
Comment (evykassirer) [inline lib/factor/ConstantFactors.js]: it'd be nice if we could pass in an argument of if we want negative factors or not - cause there'll be cases where we won't want them right?

File: lib/factor/factorQuadratic.js [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+const checks = require('../checks');
+const ConstantFactors = require('./ConstantFactors');
+const ChangeTypes = require('../ChangeTypes');
+const flatten = require('../../lib/util/flattenOperands');
+const Negative = require('../Negative');
+const Node = require('../node');
+
+// Given a node, will check if it's in the form of a quadratic equation
+// `ax^2 + bx + c`, and
+// if it is, will factor it using on of the following rules:
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: one of

File: lib/factor/factorQuadratic.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+//    - Difference of squares e.g. x^2 - 4 -> (x+2)(x-2)
+//    - Perfect square e.g. x^2 + 2x + 1 -> (x+1)^2
+//    - Sum/product rule e.g. x^2 + 3x + 2 -> (x+1)(x+2)
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: love these examples 😍

File: lib/factor/factorQuadratic.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    }
+  }
+
+  // TODO: quadratic formula
+  // a(x - (-b + sqrt(b^2 - 4ac)) / 2a)(x - (-b - sqrt(b^2 - 4ac)) / 2a)
+
+  return Node.Status.noChange(node);
+}
+
+module.exports = factorQuadratic;
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: this functionality won't exposed in mathsteps yet, right?

should we add it to the top level index.js, or wait until we figure out what a more general factor interface will look like?

File: lib/factor/factorQuadratic.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        firstTermNode = polyTerm;
+      }
+      else if (exponent.value === '1') {
+        secondTermNode = polyTerm;
+      }
+      else {
+        return Node.Status.noChange(node);
+      }
+    }
+    else {
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: when would this catch?

File: lib/factor/factorQuadratic.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+      const polyTerm = new Node.PolynomialTerm(term);
+      const exponent = polyTerm.getExponentNode(true);
+      if (exponent.value === '2') {
+        firstTermNode = polyTerm;
+      }
+      else if (exponent.value === '1') {
+        secondTermNode = polyTerm;
+      }
+      else {
+        return Node.Status.noChange(node);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: when would this catch?

File: lib/factor/factorQuadratic.js [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+      }
+    }
+    else {
+      return Node.Status.noChange(node);
+    }
+  }
+
+  if (!firstTermNode || !constantTermValue) {
+    return Node.Status.noChange(node);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: when would this catch?

File: lib/factor/factorQuadratic.js [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  const symbol = firstTermNode.getSymbolNode();
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: ooohhh in the quadratic check can you make sure they have the same symbol? and add a test for that too

File: lib/factor/factorQuadratic.js [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+  const firstTermCoeffRootValue = Math.sqrt(Math.abs(firstTermCoeffValue));
+  const firstTermCoeffRootNode = Node.Creator.constant(firstTermCoeffRootValue);
+  let constantTermRootValue = Math.sqrt(Math.abs(constantTermValue));
+  let constantTermRootNode = Node.Creator.constant(constantTermRootValue);
+
+  // check if difference of squares: are a and c squares and there is no b
+  if (!secondTermNode) {
+    // must be a difference of squares
+    if (constantTermValue < 0 &&
+        firstTermCoeffRootValue % 1 === 0 &&
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: can you make a helper function to make it clear that you're checking to see if it's an integer?

or you could use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger

File: lib/factor/factorQuadratic.js [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  if (!firstTermNode || !constantTermValue) {
+    return Node.Status.noChange(node);
+  }
+
+  const symbol = firstTermNode.getSymbolNode();
+  const firstTermCoeffValue = firstTermNode.getCoeffValue();
+  const firstTermCoeffRootValue = Math.sqrt(Math.abs(firstTermCoeffValue));
+  const firstTermCoeffRootNode = Node.Creator.constant(firstTermCoeffRootValue);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: this name is probably ok? but might be confusing

at first I thought it was an nthroot fucntion node

File: lib/factor/factorQuadratic.js [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+  let constantTermRootValue = Math.sqrt(Math.abs(constantTermValue));
+  let constantTermRootNode = Node.Creator.constant(constantTermRootValue);
+
+  // check if difference of squares: are a and c squares and there is no b
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: ... and c is negative

File: lib/factor/factorQuadratic.js [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+  if (!secondTermNode) {
+    // must be a difference of squares
+    if (constantTermValue < 0 &&
+        firstTermCoeffRootValue % 1 === 0 &&
+        constantTermRootValue % 1 === 0) {
+      const polyTerm = Node.Creator.polynomialTerm(symbol, null, firstTermCoeffRootNode);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: can you split this line into 2 lines? (and everywhere else you have a line like this)

File: lib/factor/factorQuadratic.js [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+const Node = require('../node');
+
+// Given a node, will check if it's in the form of a quadratic equation
+// `ax^2 + bx + c`, and
+// if it is, will factor it using on of the following rules:
+//    - Difference of squares e.g. x^2 - 4 -> (x+2)(x-2)
+//    - Perfect square e.g. x^2 + 2x + 1 -> (x+1)^2
+//    - Sum/product rule e.g. x^2 + 3x + 2 -> (x+1)(x+2)
+//    - TODO: quadratic formula
+function factorQuadratic(node) {
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: it'd be nice if this was split into separate function for different cases, but I also see how they all share common variables, so maybe this is best?

File: lib/factor/factorQuadratic.js [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+  }
+  else {
+    // check if perfect square: are a and c squares and b = 2*sqrt(a)*sqrt(c)
+
+    // if the second term is negative, then the constant in the parens is subtracted
+    // i.e. x^2 -2x + 1 -> (x-1)^2
+    const secondTermCoeffValue = secondTermNode.getCoeffValue();
+    if (secondTermCoeffValue < 0) {
+      constantTermRootValue = constantTermRootValue * -1;
+      constantTermRootNode = Negative.negate(constantTermRootNode);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: alternatively, you can just make `constantTermRootNode` at the times you need it instead of at the beginning

not sure if that's better or not

I have a slight preference for this because it becomes more obvious why you need it

File: test/factor/factorQuadratic.test.js [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+    ['x^2 + 2x + 1', '(x + 1)^2'],
+    ['x^2 - 2x + 1', '(x - 1)^2'],
+    ['x^2 + 3x + 2', '(x + 1)(x + 2)'],
+    ['x^2 - 3x + 2', '(x - 1)(x - 2)'],
+    ['x^2 + x - 2', '(x - 1)(x + 2)'],
+    ['x^2 + 4', 'x^2 + 4'],
+    ['x^2 + 4x + 1', 'x^2 + 4x + 1'],
+  ];
+  tests.forEach(t => testFactorQuadratic(t[0], t[1]));
+});
```
Comment (evykassirer) [inline test/factor/factorQuadratic.test.js]: can you add some tests for things that shouldn't factor yet? to make sure we're not treating things as difference of square when they're not, etc.

I'm especially cautious about stuff with signs e.g. `-x^2 - 2x - 1`

File: lib/factor/factorQuadratic.js [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    const perfectProduct = 2 * firstTermCoeffRootValue * constantTermRootValue;
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: some more comments describing how these variables fit into things might be helpful

I understood what was going on, but had to read over the code a couple times to get it

File: lib/factor/factorQuadratic.js [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+      const newNode = Node.Creator.operator('^', [paren, exponent]);
+      return Node.Status.nodeChanged(
+        ChangeTypes.FACTOR_PERFECT_SQUARE, node, newNode);
+    }
+    else if (firstTermCoeffValue === 1) {
+      // try sum/product rule: find a factor pair of c that adds up to b
+      const factorPairs = ConstantFactors.getFactorPairs(constantTermValue, true);
+      for (const pair of factorPairs) {
+        if (pair[0] + pair[1] === secondTermCoeffValue) {
+          const polyTerm = Node.Creator.polynomialTerm(symbol, null, firstTermCoeffRootNode);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: we know that the coeff is 1, so the polyTerm is just the symbol right?

File: lib/checks/isQuadratic.js (hunk 5 above)
Comment (aelnaiem) [inline lib/checks/isQuadratic.js]: Hmmmmmm. Interesting. You're right :)

File: lib/factor/ConstantFactors.js (hunk 10 above)
Comment (aelnaiem) [inline lib/factor/ConstantFactors.js]: Yea I agree and I thought about it but for now I think it's better to wait until that case arises. That seems like a better approach in general

File: lib/factor/factorQuadratic.js (hunk 15 above)
Comment (aelnaiem) [inline lib/factor/factorQuadratic.js]: It shouldn't.

File: lib/factor/factorQuadratic.js (hunk 14 above)
Comment (aelnaiem) [inline lib/factor/factorQuadratic.js]: It shouldn't.

File: lib/factor/factorQuadratic.js (hunk 16 above)
Comment (aelnaiem) [inline lib/factor/factorQuadratic.js]: It shouldn't.

File: lib/factor/factorQuadratic.js (hunk 17 above)
Comment (aelnaiem) [inline lib/factor/factorQuadratic.js]: Nice!

File: lib/factor/factorQuadratic.js (hunk 19 above)
Comment (aelnaiem) [inline lib/factor/factorQuadratic.js]: Do you have a better name in mind? I agree that the naming is annoying.

File: lib/factor/ConstantFactors.js (hunk 10 above)
Comment (evykassirer) [inline lib/factor/ConstantFactors.js]: weren't we using this factor functionality for something else? this still works there with this change?

File: lib/factor/factorQuadratic.js (hunk 15 above)
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: can we take em out then?

File: lib/factor/factorQuadratic.js (hunk 19 above)
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: I think if you wait until you need them to create/name them, it'll be more obvious what they do because the context

File: lib/factor/factorQuadratic.js (hunk 22 above)
Comment (aelnaiem) [inline lib/factor/factorQuadratic.js]: Did it :) although each function takes a bunch of parameters (a, b and c of the quadratic)

File: lib/factor/ConstantFactors.js (hunk 10 above)
Comment (aelnaiem) [inline lib/factor/ConstantFactors.js]: I actually pre-emptively added this function when adding other factoring functionality. Which caused me some confusion and which is why I think it's best to wait for the case to arise now hahaha

Comment (evykassirer) [inline lib/factor/ConstantFactors.js]: ahhh okay - good to know XD

File: lib/simplifyExpression/collectAndCombineSearch/addLikeTerms.js [hunk 27]
```diff
@@ -120,6 +121,7 @@ function addNegativeOneCoefficient(node) {
         polyTerm.getSymbolNode(),
         polyTerm.getExponentNode(),
         polyTerm.getCoeffNode(),
+        false,
```
Comment (evykassirer) [inline lib/simplifyExpression/collectAndCombineSearch/addLikeTerms.js]: add a comment for what this means maybe

File: lib/simplifyExpression/collectAndCombineSearch/addLikeTerms.js [hunk 28]
```diff
@@ -84,7 +84,8 @@ function addPositiveOneCoefficient(node) {
       newNode.args[i] = Node.Creator.polynomialTerm(
         polyTerm.getSymbolNode(),
         polyTerm.getExponentNode(),
-        Node.Creator.constant(1));
+        Node.Creator.constant(1),
+        true);
```
Comment (evykassirer) [inline lib/simplifyExpression/collectAndCombineSearch/addLikeTerms.js]: maybe add comment for what true means

File: lib/factor/factorQuadratic.js [hunk 29]
```diff
@@ -44,84 +44,187 @@ function factorQuadratic(node) {
...
     }
   }
 
-  // TODO: quadratic formula
-  // a(x - (-b + sqrt(b^2 - 4ac)) / 2a)(x - (-b - sqrt(b^2 - 4ac)) / 2a)
-
   return Node.Status.noChange(node);
 }
 
+
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: extra line

File: lib/node/Creator.js [hunk 30]
```diff
@@ -48,12 +48,12 @@ const NodeCreator = {
   // exponent might be null, which means there's no exponent node.
   // similarly, coefficient might be null, which means there's no coefficient
   // the symbol node can never be null.
-  polynomialTerm (symbol, exponent, coeff, coeffIsNegOne=false) {
+  polynomialTerm (symbol, exponent, coeff, coeffIsOne=false, coeffIsNegOne=false) {
```
Comment (evykassirer) [inline lib/node/Creator.js]: I wonder if we could combine `coeffIsOne` and `coeffIsNegOne` to a shared variable... something about explicit coefficients

or is that more confusing?

File: lib/checks/isQuadratic.js [hunk 31]
```diff
@@ -12,13 +12,14 @@ function isQuadratic(node) {
     return false;
   }
 
+  // get the number of symbols in the expression and ensure there's only one
```
Comment (evykassirer) [inline lib/checks/isQuadratic.js]: sorry I'm being really picky now so you don't have to change this but

I think this comment is pretty close to what the code is doing, I meant more like 
"make sure only one symbol appears in the expression, i.e. just x and not both x and y"

the point of confusion for people could be that it's a _set_ (which you show in the var name which is helpful, so it's probably fine) and there can be multiple symbol terms (multiple symbols) as long as they're of the same symbol name

File: lib/factor/factorQuadratic.js [hunk 32]
```diff
@@ -153,61 +144,57 @@ function factorDifferenceOfSquares(node, symbol, aValue, bValue, cValue, negate)
...
     // try sum/product rule: find a factor pair of c that adds up to b
     const factorPairs = ConstantFactors.getFactorPairs(cValue, true);
     for (const pair of factorPairs) {
       if (pair[0] + pair[1] === bValue) {
-        const aRootValue = Math.sqrt(Math.abs(aValue));
-        const aRootNode = Node.Creator.constant(aRootValue);
-
-        const polyTerm = Node.Creator.polynomialTerm(
-          symbol, null, aRootNode);
+        const polyTerm = Node.Creator.polynomialTerm(symbol, null);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: oh wait you can just use Node.Creator.symbol

**Comment by evykassirer** [comment]
I'm worried that `2x^2 + 2x -> 2x(x + 1)` might still not work

File: lib/factor/factorQuadratic.js [hunk 33]
```diff
@@ -153,61 +144,57 @@ function factorDifferenceOfSquares(node, symbol, aValue, bValue, cValue, negate)
...
-      const newNode = Node.Creator.operator('^', [paren, exponent]);
-      return Node.Status.nodeChanged(
-        ChangeTypes.FACTOR_PERFECT_SQUARE, node, newNode);
+    const polyTerm = Node.Creator.polynomialTerm(symbol, null, aRootNode);
+    let paren = Node.Creator.parenthesis(
+      Node.Creator.operator('+', [polyTerm, cRootNode]));
+    const exponent = Node.Creator.constant(2);
+
+    if (negate) {
+      paren = Negative.negate(paren);
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: if you negate before doing the ^2, won't the negating cancel out?

ie `(-(x+2))^2`

I see your test passes though, which confuses me

**Comment by evykassirer** [comment]
still missing tests for `factorSymbol` with negate - I think there might be bugs there

File: lib/factor/factorQuadratic.js [hunk 34]
```diff
@@ -56,35 +69,13 @@ function factorQuadratic(node) {
     cValue = -cValue;
```
Comment (evykassirer) [inline lib/factor/factorQuadratic.js]: these should default to 0 above, or else negating doesn't really make sense in the null case

**Comment by evykassirer** [comment]
added 6 comments, mostly just stuff you missed from the last code review

lemme know if there's anything I can do to make it easier for you to see and check off my comments to make sure you address them all (I know github doesn't really do a good job of making it easy)

**Comment by evykassirer** [comment]
👍
